### ISSUE TITLE
Panic on pathological input `𛲟<td:^𛲟`

### ISSUE THREAD

**Issue by oberien** [raphlinus/pulldown-cmark#283]
Executing `echo -ne "\xf0\x9b\xb2\x9f<td:^\xf0\x9b\xb2\x9f" | RUST_BACKTRACE=1 cargo run` (with the hex sequence being U+1BC9F encoded with UTF-8) results in the following panic backtrace:

```rust
thread 'main' panicked at 'byte index 12 is not a char boundary; it is inside '𛲟' (bytes 9..13) of `𛲟<td:^𛲟`', src/libcore/str/mod.rs:2027:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: rust_begin_unwind
             at src/libstd/panicking.rs:308
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::str::slice_error_fail
             at src/libcore/str/mod.rs:0
   9: core::str::traits::<impl core::slice::SliceIndex<str> for core::ops::range::Range<usize>>::index::{{closure}}
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libcore/str/mod.rs:1751
  10: core::option::Option<T>::unwrap_or_else
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libcore/option.rs:386
  11: core::str::traits::<impl core::slice::SliceIndex<str> for core::ops::range::Range<usize>>::index
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libcore/str/mod.rs:1751
  12: core::str::traits::<impl core::ops::index::Index<I> for str>::index
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libcore/str/mod.rs:1616
  13: pulldown_cmark::parse::scan_uri
             at src/parse.rs:1883
  14: pulldown_cmark::parse::scan_autolink
             at src/parse.rs:1830
  15: pulldown_cmark::parse::Parser::handle_inline_pass1
             at src/parse.rs:2208
  16: pulldown_cmark::parse::Parser::handle_inline
             at src/parse.rs:2188
  17: <pulldown_cmark::parse::Parser as core::iter::traits::iterator::Iterator>::next
             at src/parse.rs:2806
  18: <&mut I as core::iter::traits::iterator::Iterator>::next
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libcore/iter/traits/iterator.rs:2693
  19: pulldown_cmark::html::HtmlWriter<I,W>::run
             at ./src/html.rs:144
  20: pulldown_cmark::html::write_html
             at ./src/html.rs:522
  21: pulldown_cmark::main
             at src/main.rs:260
  22: std::rt::lang_start::{{closure}}
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libstd/rt.rs:64
  23: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  24: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  25: std::rt::lang_start_internal
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  26: std::rt::lang_start
             at /rustc/befeeb7c083c9e5edddc86563cc461185d897a13/src/libstd/rt.rs:64
  27: main
  28: __libc_start_main
  29: _start
```

**Comment by marcusklaas** [comment]
Good catch! Did you find this using a fuzzer?

Incidentally, this is also fixed by https://github.com/raphlinus/pulldown-cmark/pull/281 it seems ^^

**Comment by oberien** [comment]
I tested it against your marcusklaas/remove-inline-scanner branch and the panic doesn't happen there. You can write `Fix #283` in the toplevel comment of #281 to have this issue auto-closed once the PR is merged.

I didn't find it using a fuzzer directly, I'm currently trying to tackle #257. More on that later :)
